californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
I tried to understand better all of this reading the code, but I'm afraid I'm yet more lost.
Reading the javadoc of currentRequest :
```java
/**
* Returns the current request block. If a request is not being sent
* blockwise, the whole request counts as a single block and this method
* returns the same request as {@link #getRequest()}. Call getRequest() to
 * access the assembled request.
 * 
 * @return the current request block
*/
public Response getCurrentResponse()
```
and 
```java
/**
 * This method is only needed when the same {@link Exchange} instance uses
 * different tokens during its lifetime, e.g., when using a different token
 * for retrieving the rest of a blockwise notification (when not altered,
 * Californium reuses the same token for this).
 * See {@link BlockwiseLayer} for an example use case.
 */
public void completeCurrentRequest()
```
and knowing that blockwiselayer should be the only class which change the currentRequest (changing the orginal one by the current block request), I still feel that the cleaning should be initiated by it. 

But as I explained I'm not sure to see how all of this should work, I just want to let you know that  adding this cleaning step confusing me a bit more.